skbio/diversity/_block.py

```diff
@@ -0,0 +1,280 @@
+
+    block_kwargs['counts'] = counts_block
+    block_kwargs['ids'] = ids_to_keep
+
+    if 'tree' in kwargs and 'otu_ids' in kwargs:
+        block_kwargs['otu_ids'] = np.asarray(kwargs['otu_ids'])[nonzero_cols]
+        block_kwargs['tree'] = kwargs['tree'].shear(block_kwargs['otu_ids'])
+
+    return block_kwargs
+
+
+def _pairs_to_compute(rids, cids):
+    """Determine the pairs of samples to compute distances between
+
+    Parameters
+    ----------
+    rids : Iterable
+        The row IDs in the partial pairwise computation.
+    cids : Iterable
+        The column IDs in the partial pairwise computation.
+
+    Raises
+    ------
+    ValueError
+        When determining ID pairs for blocks that fall outside of the diagonal
+        of the resulting distance matrix, if a pair corresponds to the lower
+        triangle, complain loudly.
+
+    Returns
+    -------
+    list of tuple
+        The ID pairs to compute distances between.
+    """
+    # if identical, gather the upper triangle
+    if len(rids) == len(cids) and (rids == cids).all():
+        return [(i, j) for idx, i in enumerate(rids) for j in rids[idx+1:]]
+
+    # otherwise, grab pairwise combinations disregarding the diagonal
+    else:
+        if set(rids).intersection(set(cids)):
+            raise ValueError("Attempting to compute a lower triangle")
+        return [(i, j) for i in rids for j in cids if i != j]
+
+
+def _block_kwargs(**kwargs):
+    """Construct arguments describing a block to compute
+
+    Returns
+    -------
+    dict
+        The parameters for the block of the distance matrix to compute.
+    """
+    for row_ids, col_ids in _generate_id_blocks(kwargs['ids'], kwargs['k']):
+        id_pairs = _pairs_to_compute(row_ids, col_ids)
+        if id_pairs:
+            kw = kwargs.copy()
+            kw['id_pairs'] = id_pairs
+            kw['row_ids'] = row_ids
+            kw['col_ids'] = col_ids
+
+            yield kw
+
+
+def _block_compute(**kwargs):
+    """Compute a block within the resulting distance matrix
+
+    Notes
+    -----
+    This method encapsulates the two expensive operations to perform for each
+    block, namely, the "shearing" of the phylogenetic tree to correspond to
+    only the OTUs of interest, and the actual beta diversity calculations.
+
+    Returns
+    -------
+    DistanceMatrix
+    """
+    block_kw = _block_party(**kwargs)
+
+    return partial_beta_diversity(**block_kw)
+
+
+def _map(func, kw_gen):
+    """Map a function over arguments
+
+    Note
+    ----
+    builtin map does not allow for mapping with kwargs.
+
+    """
+    for kwargs in kw_gen:
+        yield func(**kwargs)
+
+
+def _reduce(blocks):
+    """Reduce an iterable of partial distance matrices into a full matrix"""
+    all_blocks = list(blocks)
+
+    # Determine the maximum integer ID observed in the blocks. There exists a
+    # 1-1 mapping between the integer ID and a sample ID. We increment by 1
```

Comment:
There are different ways a developer may choose to do a reduction, the implementation of which may be dependent on a parallel framework. I do not expect users to override it, but I expect developers to as necessary. In practice, the overhead for reducing partial distance matrices is very small relative to the walltime expended in mapping, so at least for the sized problems I've used this on, I haven't seen a benefit in redoing this method. That being said, there is room for a parallel reduction if someone so decided to implement it.